compiler/src/dotty/tools/dotc/interpreter/Interpreter.scala

```diff
@@ -0,0 +1,129 @@
+package dotty.tools.dotc
+package interpreter
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.core.Constants._
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.Decorators._
+import dotty.tools.dotc.core.Symbols._
+
+import scala.reflect.ClassTag
+import java.net.URLClassLoader
+
+/** Tree interpreter that can interpret
+ *   * Literal constants
+ *   * Calls to static methods
+ *   * New objects with explicit `new` keyword
+ *   * Quoted code
+ *
+ *  The interpreter assumes that all calls in the trees are to code that was
+ *  previously compiled and is present in the classpath of the current context.
+ */
+class Interpreter(implicit ctx: Context) {
+  import tpd._
+
+  private[this] val classLoader = {
+    val urls = ctx.settings.classpath.value.split(':').map(cp => java.nio.file.Paths.get(cp).toUri.toURL)
+    new URLClassLoader(urls, getClass.getClassLoader)
+  }
+
+  /** Returns the interpreted result of interpreting the code represented by the tree.
+   *  Return Some of the result or None if some error happen during the interpretation.
+   */
+  def interpretTree[T](tree: Tree)(implicit ct: ClassTag[T]): Option[T] = {
+    try {
+      interpretTreeImpl(tree) match {
+        case obj: T => Some(obj)
+        case obj =>
+          ctx.error(s"Interpreted tree returned a result of an unexpected type. Expected ${ct.runtimeClass} but was ${obj.getClass}", tree.pos)
+          throw new StopInterpretation
```

Comment:
just return `None` here.

Reply:
We probably should upgrade that to a full type tag, or whatever our equivalent quoted structure is. But that's future work.